### ISSUE TITLE
Delete Unloaded Mod World Data

### ISSUE THREAD

**Issue by JavidPack** [bluemagic123/tModLoader#302]
Sometimes, a user might want to completely clear mod data from a World from a particular mod.

One reason might be the mod data is preventing the world from loading properly. Another might be that the user no longer wants the unloaded pink tiles left behind in a world when a mod is unloaded.

I propose a menu to be available from the World Select menu, this menu, called "More" or "Tools" or "Advanced", could list many buttons that would do specific things. In terms of the above suggestions, one could be "Delete unloaded data". This tool would delete all ModWorld data that belongs to currently unloaded mods. 

Another tool could be "Delete unloaded tiles", this tool would simply delete pink/unloaded tiles.

A more advanced implementation might allow selecting which mod to delete mod data from.

**Comment by Jofairden** [comment]
Seems useful. We could generify this enough, buttons to remove any type of modded data: items, projectiles, npcs.. anything instanced basically. Unless we can think of arguments against that

**Comment by Jofairden** [comment]
@JavidPack Coming back to this. Would it be useful if modders can specify a 'fallback' type that the tiles will be replaced with if deleted? It can be stored in MysteryTile data.

**Comment by Chicken-Bones** [comment]
I'd really like to do a more thorough investigation on this. Leftover data should never corrupt worlds. Maybe people want to delete all the unloaded blocks/items etc, which is a useful feature.

I suppose if a mod updates with poor backwards compatibility, you might want to unload said mod, wipe it from the world, and then reload. I still think, this can be done just with a 'wipe unloaded data' button.

**Comment by GoodPro712** [comment]
I have a command that removed unloaded **tiles**

**Comment by Jofairden** [comment]
I think for 1.4 it would be good to have a feature so you can just completely remove unloaded data from player and world files.

**Comment by Solxanich** [comment]
Hmm. I'm thinking of tackling this next, as its related with the placeholders walls and chests request I worked on. 
To confirm the request proper: 
"For Unloaded World, aim for a method like destroyUnloadedThisModData which removes all placeholders belonging to the mod, and all related unloaded mod data.   Nice to haves for this includes a way to set 'fallbacks' for destroyed placeholders, and being able to specify particular subsets of data to wipe (tiles, items, etc.). Method can be called from World Selection UI via appropriate popup display. Method should not corrupt the save.
Likewise implement for UnloadedPlayer, being called from Character Selection UI."

**Comment by bvcbvc** [comment]
perhaps along with the deletion, perhaps another option to replace unloaded content, like with dirt or something

**Comment by Chicken-Bones** [comment]
Should be possible to show a nice little UI with a tree view for each unloaded mod and the content that will be deleted. 

For tiles/walls we can probably replace the entry with a special one like "Terraria/Tile" and use the fallback id on load.

**Comment by bvcbvc** [comment]
> Should be possible to show a nice little UI with a tree view for each unloaded mod and the content that will be deleted.
> 
> For tiles/walls we can probably replace the entry with a special one like "Terraria/Tile" and use the fallback id on load.
perhaps even being able to choose what is replaced for those that dont have fallback

**Comment by Solxanich** [comment]
> For tiles/walls we can probably replace the entry with a special one like "Terraria/Tile" and use the fallback id on load.
Yes we can use the ModBlockType modder defineable property vanillaFallbackOnDeletion to do it. Somehow forgot to reimplement in that last batch rewrite on Load.
Theres probably an if CanPurge else I can toss somewhere in the rewrite; I suspect just swapping the loadType on the bool, and checking CanPurge with IsUnloaded before indexing should work.

**Comment by TwistedAkai** [comment]
> > Should be possible to show a nice little UI with a tree view for each unloaded mod and the content that will be deleted.
> > For tiles/walls we can probably replace the entry with a special one like "Terraria/Tile"
> > use the fallback id on load.
> 
> perhaps even being able to choose what is replaced for those that dont have fallback or being able to choose in general

I know there hasn't been anything new on this issue in awhile, but I saw this and got to thinking. If it hasn't already been done, maybe have a default fallback of the biome's 'dirt' or 'stone', or some randomly selected combination of the two? Assuming it's not too much added work to determine the biome for each tile when replacing them?

**Comment by JavidPack** [comment]
I think the main concerns of this issue are resolved in #4325. I don't think we need to worry about unloaded ModSystem data, but if it ever comes up again we can start a new issue for that.
